File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool dfs(unordered_map<int, vector<int>>& graph, int& node, int& destination, vector<bool>& visited) {
        if (node==destination) return true;
        if (visited[node]==0) {
            visited[node]=1;
            for (auto neighbor : graph[node]) {
                if (dfs(graph, neighbor, destination, visited)) return true;
            }
        }
        return false;
    }

    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        if (edges.size()==0) return true;
        unordered_map<int, vector<int>> graph;
        for (auto& e:edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }
        vector<bool> visited(n, 0);
        return dfs(graph, source, destination, visited);
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(list)
        for e in edges: # bi-directional graph
            graph[e[0]].append(e[1])
            graph[e[1]].append(e[0])

        visited = [False for i in range(n)]
        def dfs(u, v):
            if u == v:
                return True
            if visited[u] == False:
                visited[u] = True
                for n in graph[u]:
                    if dfs(n, v): 
                        return True
            return False
        return dfs(source, destination)